### PR TITLE
Error Term on Utility Evalution

### LINKED ISSUE #257: [PoC] Objective Function Evaluation (U - dU_t + burntFees)
## Discussion and Justification for Objective Function 

After several iterations of solution ranking system, the PoC/MVP Smart contract will implement the following objective function:

```
TotalUtility - TotalDisregardedUtility (touched orders) + BurntFees
```

where 

### 1. **TotalUtility** 
is the sum over all each order's utility, given by

```py
U_order = [(execBuyAmt * order.sellAmt - execSellAmount * order.buyAmt) * price.buyToken] / order.sellAmt
```
simplified and evaluated by the smart contract (using integer division) as
```py
U_order = [execBuyAmt - (execSellAmount * order.buyAmt) / order.sellAmt] * price.buyToken
```

### 2. **DisregardedUtility** (of an order) 
is a measure of how much of the order could have been executed that was not. This measure is non-zero exactly when the uniform clearing price was within the order's specified limit price, but not all of the order's sell volume was used in auction settlement. In our case, the smart contract can only evaluate/verify the disregarded utility of touched orders (since otherwise, the computation would easily exceed block-gas-limits). Heuristically, `|dU|` is the utility of the untouched sellAmount (`execSellAmt - requestedSellAmt`) and the evaluation used in the smart contract is given as 

```py
|disregardedUtility| = (limitTerm * leftoverSellAmount) / order.sellAmount
```
where 
```py
limitTerm = price.SellToken * order.sellAmt - order.buyAmt * price.buyToken
```
and
```
leftoverSellAmount = order.sellAmt - execSellAmt
```

_Note that_ `limitTerm` is always non-negative, since respected limit prices are enforced by the smart contract, and represents the amount to which the order's limit price was respected. Furthermore, `dU` evaluates to zero when limit prices are met, so to say that any fraction of an order executed exactly your limit price yields no disregarded utility, in the same sense that it yields no utility.

At the point (within the contract) of `dU` evaluation balances and orders have all been updated so that 
`sellAmount - execSellAmt == order.remainingAmount` and, for correctness, we take the minimum of this with the user's token balance.

This means that disregarded utility is not double counted if, for example, the user had two orders like `A for B` and `B for C` which were both touched ... @josojo and/or @fleupold  could you perhaps elaborate on this statement?

### 3. **BurntFees** 
is introduced for two reasons:

- So that in scenarios where all orders limit price's are met exactly, the contract can distinguish this as better than the trivial solution where no trade execution takes place.
- So that direct trades are preferred by this metric (when possible) over longer rings.

## Implementation in steps

This objective evaluation will be implemented in two major steps 

1. Introducing the `utility evaluation` as the primary objective (in #236)

2. Introducing the `disregarded utility` and `burntFees` terms (in #229) 

3. Gas-estimation branch (#231) demonstrating the amount of gas used in on solution submission and how many orders can practically be touched if there are to be competing solutions in batches. 

4. Finally, as a result of gas estimations, it was decided that **solution submission should not exceed half the block-gas limit** at a healthy 4.5Million gas for to successfully submit a competing solution.  Thus, another PR #256 introduces a constraint on the maximum number of touched orders per batch at 25 orders.

### PR DESCRIPTION
**Pull request by bh2smith**
Addresses Error introduced by integer division on utility evaluation (closes #257)

### CODE REVIEW

File: contracts/StablecoinConverter.sol [hunk 1]
```diff
@@ -241,15 +241,18 @@ contract StablecoinConverter is EpochTokenLocker {
...
             execBuy,
             currentPrices[order.buyToken],
             currentPrices[order.sellToken]
-        );
-        return uint128(
-            execBuy.sub(execSell.mul(order.priceNumerator)
-                .div(order.priceDenominator)).mul(currentPrices[order.buyToken])
-        );
+        ).mul(order.priceNumerator);
+        uint256 essentialUtility = execBuy.sub(execSellTimesBuy.div(order.priceDenominator)).mul(currentPrices[order.buyToken]);
```
Comment (fleupold) [inline contracts/StablecoinConverter.sol]: not a big fan of the word essential. Maybe just utility or roundedUtility?

**Review by fleupold — COMMENTED** [review]
What's the impact on gas? I see you posted results in the gas estimation issue, but can you summarise it here?

**Comment by bh2smith** [comment]
> What's the impact on gas? I see you posted results in the gas estimation issue, but can you summarise it here?


No major impact on Gas... below are the relevant costs:

## Cost for  Submitting a Competing Solution

20 GWei with ETH @ ~183 USD

| num Orders   | Gas Used (Million) | USD Cost |
| ------------- |:-------------:| -----:|
| 10 | 2.0   | $7.24 |
| 20 | 3.8 |  $13.76 |
| 24 | 4.6 |  $16.65 |
| 30 | 5.6 |  $20.27 |
| 40 | gas limit exceeded | XXX |

**Comment by fleupold** [comment]
> No major impact on Gas... below are the relevant costs:

Thank you! Could you also share the table before the change (in particular I'm interested in how much it increased - basically quantifying how much is "no major impact")

**Comment by bh2smith** [comment]
> Could you also share the table before the change

Unfortunately, the best estimations I can provide from before are likely very outdated. The originals came from before the utility function was evaluated correctly here:

| num Orders   | Gas Used (Million) | USD Cost |
| ------------- |:-------------:| -----:|
| 10 | 1.9   | $6.54 |
| 20 | 3.58 |  $12.32 |
| 40 | gas limit exceeded | XXX |


I should probably update the issue description with the new numbers.

**Comment by fleupold** [comment]
> Unfortunately, the best estimations I can provide from before are likely very outdated. 

Would you be able to rerun the estimation when commenting out the error term calculation? How do you know the difference is not major?

**Comment by bh2smith** [comment]
> Would you be able to rerun the estimation when commenting out the error term calculation?

Very happy to run the tests again without the error term. Below are the exact estimations of gas consumption for all the tests included in our profiler: 

## Without Error Term

```bash
  Contract: StablecoinConverter
    submit single solutions
      ✓ filling 10 orders (3116ms, 2488547 gas)
      ✓ filling 20 orders (5342ms, 4547618 gas)
      ✓ filling 40 orders (12512ms, 8667434 gas)
      ✓ fails with out of gas when trying to fill 80 orders (30707ms, 12869877 gas)
    submit competing solutions
      ✓ submits 1 competing with 10 touched (5109ms, 4509435 gas)
      ✓ submits 2 competing with 10 touched (7624ms, 6530323 gas)
      ✓ submits 1 competing with 20 touched (9721ms, 8368689 gas)
      ✓ submits 2 competing with 20 touched (13548ms, 12189760 gas)
      ✓ submits 1 competing with 24 touched (10991ms, 9913168 gas)
      ✓ submits 2 competing with 24 touched (16415ms, 14454848 gas)
      ✓ submits 1 competing with 30 touched (14330ms, 12230407 gas)
      ✓ submits 2 competing with 30 touched (20416ms, 17853418 gas)
      ✓ fails with out of gas on competing solution with 40 touched (35577ms, 15389409 gas)
```
## With Error Term

```bash
  Contract: StablecoinConverter
    submit single solutions
      ✓ filling 10 orders (2619ms, 2497127 gas)
      ✓ filling 20 orders (5050ms, 4564778 gas)
      ✓ filling 40 orders (11381ms, 8701754 gas)
      ✓ fails with out of gas when trying to fill 80 orders (27675ms, 12869877 gas)
    submit competing solutions
      ✓ submits 1 competing with 10 touched (4548ms, 4526595 gas)
      ✓ submits 2 competing with 10 touched (6764ms, 6556063 gas)
      ✓ submits 1 competing with 20 touched (9103ms, 8403009 gas)
      ✓ submits 2 competing with 20 touched (14854ms, 12241240 gas)
      ✓ submits 1 competing with 24 touched (13072ms, 9954352 gas)
      ✓ submits 2 competing with 24 touched (18808ms, 14516624 gas)
      ✓ submits 1 competing with 30 touched (13911ms, 12281887 gas)
      ✓ submits 2 competing with 30 touched (20723ms, 17930638 gas)
      ✓ fails with out of gas on competing solution with 40 touched (25840ms, 15423729 gas)
```




> How do you know the difference is not major? 

I had a feeling the difference was not major because there is only 4 additional arithmetic operations (specifically `mod`, `mul`, `div` and `sub`) of difference occurring at most the number of touched orders times.

**Review by fleupold — APPROVED** [review]
Thanks for doing the checks, to summarize: it's roughly 1k gas per order more.

**Review by josojo — APPROVED** [review]
Great suggestion from @nlordell 

Great implementation and gas estimations from @bh2smith 

Go, Awesome Team!

**Pull request merged**